### PR TITLE
CAN I Take Your Temperature?

### PR DESCRIPTION
**Pull request by AarC10**

### CODE REVIEW

File: include/HUDL/HUDL.hpp [hunk 1]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+         */
+        {
+            .Key = CO_KEY(0x1400, 0, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 3,
+        },
+        {
+            .Key = CO_KEY(0x1400, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(0),
```
Comment (mjh9585) [inline include/HUDL/HUDL.hpp]: What node are you receiving from? Right now you are receiving TPDO0 from node 0. Also to make it easier when you add the node ID, store it in a constant so it can easily be changed later.

File: include/HUDL/HUDL.hpp [hunk 2]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+         */
+        {
+            .Key = CO_KEY(0x1401, 0, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 3,
+        },
+        {
+            .Key = CO_KEY(0x1401, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(1),
```
Comment (mjh9585) [inline include/HUDL/HUDL.hpp]: Same thing here.

File: include/HUDL/HUDL.hpp [hunk 3]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+         */
+        {
+            .Key = CO_KEY(0x1600, 0, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 2,
+        },
+        {
+            .Key = CO_KEY(0x1600, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = CO_LINK(0x2100, 0, 8),
```
Comment (mjh9585) [inline include/HUDL/HUDL.hpp]: This is an 8 bit number, but the value you are linking to is 32 bit. These should be the same size. Check the sizes of all of your links.

File: include/HUDL/HUDL.hpp [hunk 4]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+            .Data = (uintptr_t) 0xFE,
+        },
+
+        /**
+         * RPDO0 mapping, determines the PDO messages to send when RPDO1 is triggered
+         * 0: The number of PDO message associated with the RPDO
+         * 1: Link to the first PDO message
+         * n: Link to the nth PDO message
+         */
+        {
```
Comment (mjh9585) [inline include/HUDL/HUDL.hpp:213]: Nitpicky thing: Add comments to your mapping entries saying what they are linked to.

File: include/HUDL/HUDL.hpp [hunk 5]
```diff
@@ -13,14 +16,19 @@ namespace HUDL {
 class HUDL {
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp:22]: The HUDL class should have a top-level comment

File: include/HUDL/HUDL.hpp [hunk 6]
```diff
@@ -13,14 +16,19 @@ namespace HUDL {
...
     /**
-    * Default Constructor for the HUDL class
-    *
-    * @param[in] reg_select is the register select pin
-    * @param[in] reset is the reset pin
-    * @param[in] cs is the chip select pin
-    * @param[in] spi is the SPI instance
-    */
-    HUDL(IO::GPIO& reg_select, IO::GPIO& reset, IO::GPIO& cs, IO::SPI& spi);
+* The node ID used to identify the device on the CAN network.
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: This spacing is incorrect. I'm not sure why the linter didn't fix this, but indent this once more and see if the linter does anything

File: include/HUDL/HUDL.hpp [hunk 7]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+     * Gets the object dictionary
+     *
+     * @return an object dictionary
+     */
+    CO_OBJ_T* getObjectDictionary();
+
+    /**
+     * Gets the size of the Object Dictionary
+     *
+     * @return uint16_t size of the Object Dictionary
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: NIT: You don't have to respecify the return type in the documentation

File: include/HUDL/HUDL.hpp [hunk 8]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+    /**
+     * Gets temperature values
+     *
+     * @return uint32_t* pointer of temperature values
+     */
+    uint32_t* getThermTemps();
+
 private:
+    /**
+     * reg_select PA_3
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp:98]: Why is this here?

File: include/HUDL/HUDL.hpp [hunk 9]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+     * reset      PB_3
+     * cs         PB_12
+     */
     DEV::LCD lcd;
+    uint16_t voltages[1] = {};
+
+    uint32_t thermTemps[5] = {0, 0, 0, 0};
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: This array only has 4 elements

File: include/HUDL/HUDL.hpp [hunk 10]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+
+    static constexpr uint16_t OBJECT_DICTIONARY_SIZE = 30;
+
+    CO_OBJ_T objectDictionary[OBJECT_DICTIONARY_SIZE + 1] = {
+        // Sync ID, defaults to 0x80
+        {CO_KEY(0x1005, 0, CO_UNSIGNED32 | CO_OBJ_D__R_), nullptr, (uintptr_t) 0x80},
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: Add a comma after `0x80`

File: src/HUDL/HUDL.cpp [hunk 11]
```diff
@@ -38,4 +37,24 @@ void HUDL::initLCD() {
...
 
+CO_OBJ_T* HUDL::getObjectDictionary() {
+    return &objectDictionary[0];
+}
+
+uint16_t HUDL::getObjectDictionarySize() const {
+    return OBJECT_DICTIONARY_SIZE;
+}
+
+uint32_t* HUDL::getThermTemps() {
```
Comment (mjmagee991) [inline src/HUDL/HUDL.cpp]: instead of exposing the thermistor and voltage values here, make a void function called something like `updateLCD()` that will later be used to regularly update the LCD. For now, just put logging statements inside updateLCD to print all the data you want to be able to see.

File: targets/HUDL/main.cpp [hunk 12]
```diff
@@ -5,6 +5,13 @@
  */
 #include <stdint.h>
 
+#include "Canopen/co_core.h"
+#include "Canopen/co_tmr.h"
+#include "EVT/dev/LED.hpp"
+#include "EVT/utils/log.hpp"
```
Comment (mjmagee991) [inline targets/HUDL/main.cpp]: These should be `<>` includes

File: targets/HUDL/main.cpp [hunk 13]
```diff
@@ -13,53 +20,180 @@
...
+ * will be passed to the EVT-core CAN interface which will in turn call this
+ * function each time a new CAN message comes in.
+ *
+ * NOTE: For this sample, every non-extended (so 11 bit CAN IDs) will be
+ * assumed to be intended to be passed as a CANopen message.
+ *
+ * @param message[in] The passed in CAN message that was read.
+ */
+
+IO::UART& uart = IO::getUART<IO::Pin::PB_10, IO::Pin::PB_11>(9600);
```
Comment (mjmagee991) [inline targets/HUDL/main.cpp]: I know you probably put this in to make testing easier and quicker, but I would much prefer we didn't have top-level variables in the main target for a board. Can you convert this to work with the logger instead?

File: targets/HUDL/main.cpp [hunk 14]
```diff
@@ -13,53 +20,180 @@
...
+    auto& hudl_spi = IO::getSPI<IO::Pin::SPI_SCK, IO::Pin::SPI_MOSI>(
         devices, deviceCount);
-    hudl_spi.configureSPI(SPI_SPEED, SPI_MODE3, SPI_MSB_FIRST);
 
-    auto board = HUDL::HUDL(reg_select, reset, cs, hudl_spi);
+    hudl_spi.configureSPI(SPI_SPEED, SPI_MODE0, SPI_MSB_FIRST);
 
-    // Setup UART
-    IO::UART& uart = IO::getUART<IO::Pin::UART_TX, IO::Pin::UART_RX>(9600);
+    auto hudl = HUDL::HUDL(regSelect, reset, hudl_spi);
```
Comment (mjmagee991) [inline targets/HUDL/main.cpp]: NIT: I'm not sure about this at all, but this declaration looks a little Java-ey to me. I think you could accomplish the same thing in a more standard C++ way with `HUDL::HUDL hudl(regSelect, reset, hudl_spi);`

File: targets/HUDL/main.cpp [hunk 15]
```diff
@@ -13,53 +20,180 @@
...
 
-        // echoes command back
-        uart.printf("\n\recho: %s\n\r", buf);
+        CONodeProcess(&canNode);
+        // Update the state of timer based events
+        COTmrService(&canNode.Tmr);
+        // Handle executing timer events that have elapsed
+        COTmrProcess(&canNode.Tmr);
+        // Wait for new data to come in
+        time::wait(1000);
```
Comment (mjmagee991) [inline targets/HUDL/main.cpp]: I believe for most boards this is 100 ms. Is there a reason you lengthened it here?

File: include/HUDL/HUDL.hpp [hunk 16]
```diff
@@ -30,37 +38,240 @@ class HUDL {
...
+    uint32_t* getThermTemps();
+
 private:
+    /**
+     * reg_select PA_3
+     * reset      PB_3
+     * cs         PB_12
+     */
     DEV::LCD lcd;
+    uint16_t voltages[1] = {};
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp:103]: I don't think you need to track the voltage data because that's not going to be useful to display on the LCD

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Looks pretty good. I just had a couple thoughts on your specific implementation

File: include/HUDL/HUDL.hpp (hunk 9 above)
Comment (mjh9585) [reply include/HUDL/HUDL.hpp]: This can also be `uint32_t thermTemps[5] = { }` to initialize it all to zero.

**Review by mjh9585 — CHANGES_REQUESTED** [review]
Just a few things to change with your object dictionary. Overall looks good.

File: include/HUDL/HUDL.hpp (hunk 8 above)
Comment (AarC10) [reply include/HUDL/HUDL.hpp:98]: I believe we wanted this commented out for the pins of the LCD we were using. We talked about having this when reviewing the code to get the LCD working in the first place

File: include/HUDL/HUDL.hpp (hunk 16 above)
Comment (AarC10) [reply include/HUDL/HUDL.hpp:103]: This was something LJ requested. In the next PR this is just a total voltage and not an array of voltages.

File: include/HUDL/HUDL.hpp (hunk 3 above)
Comment (AarC10) [reply include/HUDL/HUDL.hpp:218]: I thought this seemed off. It looks like the link sizes in the codebases for the BMS and TMS need to be fixed then

Comment (AarC10) [reply include/HUDL/HUDL.hpp]: Ok thought this seemed a little off. I just tried matching the links with the TMS and BMS object dictionaries. It seems they do the same thing

File: src/HUDL/HUDL.cpp [hunk 17]
```diff
@@ -5,14 +5,16 @@
...
 
 // clang-format off
-#include "EVT/dev/LCD.hpp"
-#include "EVT/io/GPIO.hpp"
-#include "EVT/io/SPI.hpp"
+#include <EVT/dev/LCD.hpp>
+#include <EVT/io/GPIO.hpp>
+#include <EVT/io/SPI.hpp>
+#include <EVT/utils/log.hpp>
 #include "HUDL/HUDL.hpp"
```
Comment (mjmagee991) [inline src/HUDL/HUDL.cpp]: This should also be a `<>` include

File: include/HUDL/HUDL.hpp [hunk 18]
```diff
@@ -105,18 +102,24 @@ class HUDL {
     DEV::LCD lcd;
     uint16_t voltages[1] = {};
 
-    uint32_t thermTemps[5] = {0, 0, 0, 0};
+    uint32_t thermTemps[4] = {};
 
     static constexpr uint16_t OBJECT_DICTIONARY_SIZE = 30;
+    static constexpr uintptr_t TMS_TPDO0_COB_ID = CO_COBID_TPDO_DEFAULT(0);
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: @mjh9585 should correct me if I'm wrong, but I think he meant that to get the message to receive correctly, you should use `.Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(0) + SENDER_NODE_ID,` where `SENDER_NODE_ID` is the constant you're storing, so you can use it for multiple RPDOs. I'm not sure how you were able to get the RPDOs to work without setting the COBID of the TPDO correctly, so I am worried this is going to break what you were doing before.

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Almost there

File: include/HUDL/HUDL.hpp (hunk 18 above)
Comment (mjh9585) [reply include/HUDL/HUDL.hpp]: Correct, the CO_COBID_TPDO_DEFAULT only sets the base COB-ID for a particular TPDO number. That ID then needs to be added with the NodeID that is transmitting that message. I have updated the document to more clearly reflect that. https://docs.google.com/document/d/1XJqG_xp2xKjrICqzIaXlZYilkB8anI9MclzCVAr_hfA 

Right now you are receiving PDO 0 and 1 from Node 0 which we do not use.

File: include/HUDL/HUDL.hpp [hunk 19]
```diff
@@ -30,37 +44,237 @@ class HUDL {
...
+            .Type = nullptr,
+            .Data = (uintptr_t) 0xFE,
+        },
+
+        /**
+         * RPDO0 mapping, determines the PDO messages to send when RPDO0 is triggered
+         * 0: The number of PDO message associated with the RPDO
+         * 1: Link to the first PDO message - tempThree
+         * 2: Link to the second PDO message - tempFour
+         */
```
Comment (mjh9585) [inline include/HUDL/HUDL.hpp:212]: Your comment appears to be a duplicate of RPDO map 1 and does not match the variables linked to in this map.

File: include/HUDL/HUDL.hpp (hunk 18 above)
Comment (AarC10) [reply include/HUDL/HUDL.hpp]: Ok I better understand what you guys are saying now. So the reason it works with the TMS then is because in the TMS code, the node ID is not added and its just TPDO_DEFAULT(0) and TPDO_DEFAULT(1)

Comment (AarC10) [reply include/HUDL/HUDL.hpp]: I can store the sender IDs here but it will most likely not work with the TMS until that gets fixed there

**Review by mjh9585 — CHANGES_REQUESTED** [review]
Just a few small changes.

File: include/HUDL/HUDL.hpp (hunk 18 above)
Comment (mjh9585) [reply include/HUDL/HUDL.hpp]: Yeah, for now you can test with the sender ID of 0 until that gets fixed. It looks like it will be fixed as soon as this pull request is approved: https://github.com/RIT-EVT/TMS/pull/11

File: include/HUDL/HUDL.hpp [hunk 20]
```diff
@@ -105,8 +105,8 @@ class HUDL {
     uint32_t thermTemps[4] = {};
 
     static constexpr uint16_t OBJECT_DICTIONARY_SIZE = 30;
-    static constexpr uintptr_t TMS_TPDO0_COB_ID = CO_COBID_TPDO_DEFAULT(0);
-    static constexpr uintptr_t TMS_TPDO1_COB_ID = CO_COBID_TPDO_DEFAULT(1);
+    static constexpr uintptr_t TMS_NODE_ID = 0x02;
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: The TMS's node ID is 0x08. You can find all the node IDs that the boards are supposed to have in [this doc](https://docs.google.com/spreadsheets/d/1KUt-7LYNFojA4RDHoGJRBh8pZ9A7QXzuhNFz2hbOP20/edit?usp=sharing), which is also pinned in #subgroup-firmware with the oh so helpful label "Can whatever"

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
One more thing

**Comment by mjmagee991** [comment]
[HUDL CAN](https://trello.com/c/vCpuUwvL/39-hudl-can)

**Review by mjmagee991 — APPROVED** [review]
Nice work. This has been a long time coming, but I'm really looking forward to seeing the HUDL performing next semester.

**Review by mjh9585 — APPROVED** [review]
Also looks good.
